py/mpprint: Handle %lld and similar format specifiers

stinos opened the pull request:
Just also ignore the second 'l' if one is found already.

For 64bit windows builds (both msvc and mingw-w64) the format specifier for
64bit integers is %lld or %llu for the unsigned version. These are fetched
from inttypes.h which is apparently the C99 way of working with printf/scanf
now; this just to illustrate why uPy should probably support them.

More importantly this makes all tests pass again for 64bit windows builds
which would previously fail for anything printing ranges (builtin_range/unpack1)
because they were printed as range( ld, ld )

dpgeorge commented:
Looks like we were thinking about the same thing!

The patch I just pushed actually supports %lld/%llu, and this would circumvent it.  Maybe if `__LP64__` is defined then it just skips the second 'l'.  Does that work for your case?

stinos commented:
Ah sorry I didn't notice your patch actually does the same thing. __LP64__ wouldn't work as msvc doesn't define it (nor does mingw I think), and it might brake things if it would. _WIN64 is defined for both windows ports though so that would work.

stinos committed:
py/mpprint: Fix printing of 64bit integers for 64bit windows builds

This makes all tests pass again for 64bit windows builds which would
previously fail for anything printing ranges (builtin_range/unpack1)
because they were printed as range( ld, ld ).

This is done by reusing the mp_vprintf implementation for MICROPY_OBJ_REPR_D
for 64bit windows builds (both msvc and mingw-w64) since the format specifier
used for 64bit integers is also %lld, or %llu for the unsigned version.

Note these specifiers used to be fetched from inttypes.h, which is the
C99 way of working with printf/scanf in a portable way, but mingw-w64
wants to be backwards compatible with older MS C runtimes and uses
the non-portable %I64i instead of %lld in inttypes.h, so remove the use
of said header again in mpconfig.h and define the specifiers manually.

stinos commented:
Ok this turned out to be a bit more cumbersome that I thought (reflected in updated commit message): reusing changes from your patch works just fine for msvc but not for mingw as it still uses the old non-portable %I64i format specifiers in inttypes.h. So I had to get rid of it again and just specify %lld and %llu manually in mpconfig.h..

dpgeorge commented:
Rebased and merged in 0a4eb4dbf2df34b9a2efcf55855b9db7c7132bf7, thank you!
